src/beast/beast/asio/consuming_buffers.h

```diff
@@ -0,0 +1,227 @@
+*/
+template<class Buffers,
+    class Buffer = typename Buffers::value_type>
+class consuming_buffers
+{
+    using iter_type =
+        typename Buffers::const_iterator;
+
+    static_assert(std::is_constructible<Buffer,
+        typename std::iterator_traits<iter_type>::value_type>::value,
+            "Buffer requirements not met");
+
+    Buffers bs_;
+    iter_type begin_;
+    std::size_t skip_ = 0;
+
+public:
+    using value_type = Buffer;
+
+    class const_iterator;
+
+    consuming_buffers(consuming_buffers&&) = default;
+    consuming_buffers(consuming_buffers const&) = default;
+    consuming_buffers& operator=(consuming_buffers&&) = default;
+    consuming_buffers& operator=(consuming_buffers const&) = default;
+
+    explicit
+    consuming_buffers(Buffers const& bs);
+
+    const_iterator
+    begin() const;
+
+    const_iterator
+    end() const;
+
+    /** Remove bytes from the beginning of the buffers
+
+        @param n The number of bytes to remove. If this is
+        larger than the number of bytes remaining, all the
+        bytes remaining are removed.
+    */
+    void
+    consume(std::size_t n);
+};
+
+template<class Buffers, class Buffer>
+class consuming_buffers<Buffers, Buffer>::const_iterator
+{
+    friend class consuming_buffers<Buffers, Buffer>;
+
+    using iter_type =
+        typename Buffers::const_iterator;
+
+    iter_type it_;
+    consuming_buffers const* b_;
+
+public:
+    using value_type =
+        typename std::iterator_traits<iter_type>::value_type;
+    using pointer = value_type const*;
+    using reference = value_type;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    const_iterator() = default;
+    const_iterator(const_iterator&& other) = default;
+    const_iterator(const_iterator const& other) = default;
+    const_iterator& operator=(const_iterator&& other) = default;
+    const_iterator& operator=(const_iterator const& other) = default;
+
+    bool
+    operator==(const_iterator const& other) const
+    {
+        return b_ == other.b_ && it_ == other.it_;
+    }
+
+    bool
+    operator!=(const_iterator const& other) const
+    {
+        return !(*this == other);
+    }
+
+    reference
+    operator*() const
+    {
+        if(it_ == b_->begin_)
+            return *it_ + b_->skip_;
+        return *it_;
+    }
+
+    pointer
+    operator->() const = delete;
+
+    const_iterator&
+    operator++()
+    {
+        ++it_;
+        return *this;
+    }
+
+    const_iterator
+    operator++(int)
+    {
+        auto temp = *this;
+        ++(*this);
+        return temp;
+    }
+
+    const_iterator&
+    operator--()
+    {
+        --it_;
+        return *this;
+    }
+
+    const_iterator
+    operator--(int)
+    {
+        auto temp = *this;
+        --(*this);
+        return temp;
+    }
+
+private:
+    const_iterator(consuming_buffers const& b,
+            typename Buffers::const_iterator it)
```

Comment:
Style nit:  You could use `iter_type` here instead of the wordier `typename Buffers::const_iterator`.